### PR TITLE
JS: support the "type" field on `package.json` files while extracting

### PR DESCRIPTION
**Pull request by erik-krogh**
A followup on the [`.cjs` file support](https://github.com/github/codeql/pull/4096). 

Adds support for the ["type" field from the `package.json` file](https://nodejs.org/api/esm.html#esm_package_json_type_field) in the extractor.  
The "type" field determines which module-type (CommonJS/ES2015) a  `.js` file should be treated as. 

We need this information in the extractor for e.g. [determining strict-mode](https://github.com/github/codeql/pull/4092).   
So that is why I put it into the extractor, even though it might have been easier to implement on the `.ql` side. 

I've added a `folder`->`type` cache that should hopefully eliminate any potential performance impact.

My previous `.cjs` support [PR](https://github.com/github/codeql/pull/4096) had a bug (missing `.withPlatform(Platform.NODE)`), which is why the `.trap` file has been updated by this PR.

### CODE REVIEW

File: javascript/extractor/src/com/semmle/js/extractor/ScriptExtractor.java [hunk 1]
```diff
@@ -78,4 +96,40 @@ public LoCInfo extract(TextualExtractor textualExtractor) {
...
+      return null;
+    }
+    if (!file.exists()) {
+      String result = getPackageType(folder.getParentFile());
+      packageTypeCache.put(folder, Optional.ofNullable(result));
+      return result;
+    }
+    try {
+      BufferedReader reader = new BufferedReader(new FileReader(file));
+      String result = new Gson().fromJson(reader, PackageJSON.class).type;
```
Comment (asgerf) [inline javascript/extractor/src/com/semmle/js/extractor/ScriptExtractor.java:128]: Could you add a test where `type` is present but is not a string or `null` (e.g a number). It seems like this will crash the extractor.

Also catch `JsonSyntaxException` below in case the file has a syntax error.

Comment (erik-krogh) [reply javascript/extractor/src/com/semmle/js/extractor/ScriptExtractor.java:129]: `JsonSyntaxException` seems to also catch the case where `type` is not a string.

**Review by asgerf — APPROVED** [review]
